### PR TITLE
feat: Implement profile pictures

### PR DESCRIPTION
**Pull request by ulan**
This allows a user to upload a profile picture via settings.
The profile picture is rendered in posts and comments as thumbnail.

Changes in the backend:
- add an `avatar` field to `User`.
- add an `avatar` field to `Meta`.
- add an `upload_avatar` update endpoint.
- extend existing queries to return avatar.

Changes in the frontend:
- move image helper functions to `image.ts`.
- add a profile picture field in the settings.
- show the cost of saving the settings on the button label.
- add a new `AvatarElement` helper.
- extend `UserData` to include the avatar URL.
- extend `UserLink` and `UserList` to pass user data.
- replace the user name in the header page with a new PROFILE link.
- display the profile picture in posts, comments, and profile.

### CODE REVIEW

**Comment by ulan** [comment]
@taggrx  I will rebase this once your PR is merged.

File: src/backend/env/user.rs [hunk 1]
```diff
@@ -67,6 +67,21 @@ impl UserFilter {
     }
 }
 
+/// Contains the location of the Avatar image blob.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Avatar {
+    // The canister id of the bucket canister.
+    pub bucket_id: String,
```
Comment (taggrx) [inline src/backend/env/user.rs:74]: Ouch, ouch, ouch 🙈 😂  Do you know that the current buckets implementation a pretty stupid "append only" storage and that if posts are deleted, the space is not freed up? (same would hold for avatars)

Did you think about instead simply storing these blob in the current memory, by creating an ObjectManager for these small blobs? There, both storage and removal is already supported.

File: src/frontend/src/settings.tsx [hunk 2]
```diff
@@ -70,6 +85,46 @@ export const Settings = ({ invite }: { invite?: string }) => {
...
 
+    const onAvatarChange = async (ev: any) => {
+        ev.preventDefault();
+        if (!ev.target.files[0]) {
+            return;
+        }
+        const max_blob_size_bytes =
+            window.backendCache.config.max_blob_size_bytes;
+        let blob = null;
+        for (let size of [768, 512, 384, 256, 128, 64, 32]) {
```
Comment (taggrx) [inline src/frontend/src/settings.tsx:96]: Is this for a better quality when downscaling?

**Review by xqxpx — COMMENTED** [review]
Thanks Ulan, this is super cool! 🚀 I did a super high-level review, it looks great, but let's first quickly discuss the storage question. I'm curios of you considered storing in the stable memory.

File: src/backend/env/user.rs (hunk 1 above)
Comment (ulan) [reply src/backend/env/user.rs:74]: Conceptually, on the high-level, I don't see difference between changing an avatar and posting a picture as long as the costs are paid. I don't think users would mind paying the photo fee.

If you'd like to optimize it, let's move avatars to the main storage. I like the effect that we can have a stable URL for avatars. I'll implement it after rebasing.

File: src/frontend/src/settings.tsx (hunk 2 above)
Comment (ulan) [reply src/frontend/src/settings.tsx:96]: Yeah, and also for simplicity. I was not in the mood for the binary search :D

**Comment by ulan** [comment]
Closing in favor of the suggested alternative storage.

**Pull request closed without merging**